rfcs/0007-repartitioning.md

```diff
@@ -0,0 +1,139 @@
+# Repartitioning Primitives
+
+## Motivation and Problem Statement
+
+Currently, SlateDB does not include a built-in partitioning mechanism. To scale horizontally, users must implement their own partitioning strategy and manage each partition as a separate SlateDB instance. This design is consistent with other LSM-tree–based embedded databases, where partitioning is often left to higher-level orchestration.
+
+While creating static partitions is relatively straightforward, changing the number of partitions over time (repartitioning) is more complex. Existing embedded databases like RocksDB and Badger do not provide convenient primitives to enable a simple and performant repartitioning workflow in user applications.
+
+## High-level Solution
+
+SlateDB already offers a critical feature for repartitioning: snapshots. Users can take a snapshot of the database and clone it without creating a full copy of the original data. This capability significantly reduces the overhead of duplicating data and is a strong foundation for any repartitioning strategy.
+
+To fully support a robust and efficient repartitioning process, SlateDB needs two additional primitives:
+1. Creating Partial, Non-Overlapping Clones
+2. Merging Multiple Clones
+
+### Creating Partial, Non-Overlapping Clones
+
+The ability to take a single snapshot and split it into multiple, smaller "partial" clones. Each partial clone contains only a subset of the data, allowing users to distribute different segments of the database across multiple new instances.
+
+This feature can piggyback on the existing clone mechanism - we just need the ability to hide unwanted parts of the keyrange. We can achieve this by adding additional metadata to the parent pointer of the cloned database.
+
+### Merging Multiple Clones
+
+The ability to merge multiple, non-overlapping clones back into a single database. This would enable consolidation of data when reducing the number of partitions or combining data from multiple shards into a single, cohesive database instance.
+
+## Implementation Details
+
+### KeyRange aware manifest
+
+It's crucial to be aware of the full keyrange handled by the SST file and therefore the whole database. This will allow us to effectively mask out part of the overall keyrange and ensure two databases don't overlap.
+
+We already have information about the first_key for each SST file stored in the manifest, so we simply need to add last_key to be fully aware of the contained keyrange.
+
+```rust
+// Has metadata about a SST file.
+table SsTableInfo {
+    // First key in the SST file.
+    first_key: [ubyte];
+    // Last key in the SST file.
+    last_key: [ubyte];
```

Comment:
I think this should be optional. We only need it for the L0 SSTs and the last SST in an SR. The keys in the manifest make up a good % of the manifest's overall size, so it feels worthwhile to try to avoid including this where we can.

Reply:
Hmm, the manifest size is a fair point :this: I guess we'd only need last_key for "clipped" SSTs.

We might be also able to filter out whole SSTs and in some cases starting point of the next SST might not provide enough information. I assume this will be rare and we can simply fallback to reading metadata block from S3.